crypto/evp/kdf_meth.c

```diff
@@ -0,0 +1,192 @@
+/*
+ * Copyright 2019 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <openssl/evp.h>
+#include <openssl/err.h>
+#include <openssl/core.h>
+#include <openssl/core_numbers.h>
+#include <openssl/kdf.h>
+#include "internal/evp_int.h"
+#include "internal/provider.h"
+#include "evp_locl.h"
+
+static int evp_kdf_up_ref(void *vkdf)
+{
+    EVP_KDF *kdf = (EVP_KDF *)vkdf;
+    int ref = 0;
+
+    CRYPTO_UP_REF(&kdf->refcnt, &ref, kdf->lock);
+    return 1;
+}
+
+static void evp_kdf_free(void *vkdf){
+    EVP_KDF *kdf = (EVP_KDF *)vkdf;
+    int ref = 0;
+
+    if (kdf != NULL) {
+        CRYPTO_DOWN_REF(&kdf->refcnt, &ref, kdf->lock);
+        if (ref <= 0) {
+            ossl_provider_free(kdf->prov);
+            OPENSSL_free(kdf->name);
+            CRYPTO_THREAD_lock_free(kdf->lock);
+            OPENSSL_free(kdf);
+        }
+    }
+}
+
+static void *evp_kdf_new(void)
+{
+    EVP_KDF *kdf = NULL;
+
+    if ((kdf = OPENSSL_zalloc(sizeof(*kdf))) == NULL
+        || (kdf->lock = CRYPTO_THREAD_lock_new()) == NULL) {
+        OPENSSL_free(kdf);
+        return NULL;
+    }
+    kdf->refcnt = 1;
+    return kdf;
+}
+
+static void *evp_kdf_from_dispatch(const char *name, const OSSL_DISPATCH *fns,
+                                   OSSL_PROVIDER *prov)
+{
+    EVP_KDF *kdf = NULL;
+    int fnkdfcnt = 0, fnctxcnt = 0;
+
+    if ((kdf = evp_kdf_new()) == NULL
+        || (kdf->name = OPENSSL_strdup(name)) == NULL) {
+        evp_kdf_free(kdf);
+        EVPerr(0, ERR_R_MALLOC_FAILURE);
+        return NULL;
+    }
+
+    for (; fns->function_id != 0; fns++) {
+        switch (fns->function_id) {
+        case OSSL_FUNC_KDF_NEWCTX:
+            if (kdf->newctx != NULL)
+                break;
+            kdf->newctx = OSSL_get_OP_kdf_newctx(fns);
+            fnctxcnt++;
+            break;
+        case OSSL_FUNC_KDF_DUPCTX:
+            if (kdf->dupctx != NULL)
+                break;
+            kdf->dupctx = OSSL_get_OP_kdf_dupctx(fns);
+            break;
+        case OSSL_FUNC_KDF_FREECTX:
+            if (kdf->freectx != NULL)
+                break;
+            kdf->freectx = OSSL_get_OP_kdf_freectx(fns);
+            fnctxcnt++;
+            break;
```

Comment:
I found another reason for memleaks: you don't pick up `OSSL_FUNC_KDF_RESET`
(I don't think it should be a mandatory function, so shouldn't affect `fnctxcnt`)